share/src/main/java/org/apache/kafka/server/share/persister/PersisterStateManager.java

```diff
@@ -0,0 +1,863 @@
+        }
+
+        /**
+         * The incoming requests will have the keys in the following format
+         * groupId: [
+         * topidId1: [part1, part2, part3],
+         * topicId2: [part1, part2, part3]
+         * ...
+         * ]
+         * Hence, the total number of keys would be 1 x m x n (1 is for the groupId) where m is number of topicIds
+         * and n is number of partitions specified per topicId.
+         * <p>
+         * For each RPC, we need to identify the coordinator node first.
+         * If the internal share state topic is not found in the metadata cache, when RPC is received
+         * we will need to make a FIND_COORDINATOR RPC which will have the side effect of creating the internal
+         * topic as well. If the node is found in the cache, we will use it directly.
+         *
+         * @return list of requests to send
+         */
+        @Override
+        public Collection<RequestAndCompletionHandler> generateRequests() {
+            // There are two sources for requests here:
+            // 1. A queue which will contain FIND_CORD RPCs and other non-batchable RPCs.
+            // 2. A hashMap keyed on the share coordinator nodes which may contain batched requests.
+
+            if (generateCallback != null) {
+                generateCallback.run();
+            }
+            List<RequestAndCompletionHandler> requests = new ArrayList<>();
+
+            // honor queue first as find coordinator
+            // is mandatory for batching and sending the
+            // request to correct destination node
+            if (!queue.isEmpty()) {
+                PersisterStateManagerHandler handler = queue.peek();
+                queue.poll();
+                if (handler.lookupNeeded()) {
+                    // we need to find the coordinator node
+                    Node randomNode = randomNode();
+                    if (randomNode == Node.noNode()) {
+                        log.error("Unable to find node to use for coordinator lookup.");
+                        // fatal failure, cannot retry or progress
+                        // fail the RPC
+                        handler.findCoordinatorErrorResponse(Errors.COORDINATOR_NOT_AVAILABLE, Errors.COORDINATOR_NOT_AVAILABLE.exception());
+                        return Collections.emptyList();
+                    }
+                    log.debug("Sending find coordinator RPC");
+                    return Collections.singletonList(new RequestAndCompletionHandler(
+                        time.milliseconds(),
+                        randomNode,
+                        handler.findShareCoordinatorBuilder(),
+                        handler
+                    ));
+                } else {
+                    // useful for tests and
+                    // other RPCs which might not be batchable
+                    if (!handler.isBatchable()) {
+                        requests.add(new RequestAndCompletionHandler(
```

Comment:
At the time of writing this condition is will not be used by any RPC at least for EA of share groups since both read and write are batchable. This is added for the sake of testing and completeness.